Sources/Apodini/Request/Request.swift

```diff
@@ -5,28 +5,64 @@
 //  Created by Paul Schmiedmayer on 7/12/20.
 //
 
-import Vapor
+import Foundation
+import NIO
+@_implementationOnly import Vapor
 
+/// A type representing a server response.
+public protocol Request {
+    /// parameters
+    func parameter<T: Codable>(for parameter: Parameter<T>) throws -> T
+    /// The EventLoop associated with this request
+    var eventLoop: EventLoop { get }
+    /// database
+    var database: Any? { get }
+}
+
+typealias VaporRequest = Vapor.Request
+
+extension VaporRequest: Request {
+    public func parameter<T: Codable>(for parameter: Parameter<T>) throws -> T {
+
+        switch parameter.option(for: PropertyOptionKey.http) {
+        case .path where T.self is LosslessStringConvertible:
+            guard let parameter = parameters.get(parameter.id.uuidString) as? T else {
+                throw Vapor.Abort(.badRequest)
+            }
+            return parameter
+        case .query:
+            guard let name = parameter.name else {
+                throw Vapor.Abort(.badRequest)
+            }
+            return try query.get(T.self, at: name)
+        case .body:
+            return try content.decode(T.self)
+        default:
+            throw Vapor.Abort(.badRequest)
+        }
```

Comment:
Not a big fan of the rather undescriptive error handling. Not saying this has to be resolved now, but once incorporated into the REST exporter, this should be properly described (especially the one which handles missing query parameters).
Not sure if the other two cases fall into the `badRequest` category?
Would suggest to annotate it with a `// TODO` or a `#warning` so it isn't forgotten.

Reply:
🙈 yeah this is still a bit unrefined. I'll create some custom error type for this.

Reply:
Added a dedicated error type to fix this.